### ISSUE TITLE
Using DNS2020 synthetic  no_reverb dataset compute  metrics  score is not high

### ISSUE THREAD

**Issue by zuowanbushiwo** [Rikorose/DeepFilterNet#137]
Deep Filter2  ：
SI_SDR:-25.970385364691417
STOI:0.8182306923079351
WB_PESQ:2.5978289715449017
NB_PESQ:3.2630579535166424
SI_SDR scores is very strange？

noisy：
SI_SDR:9.070872648202446
STOI:0.9151854128107001
WB_PESQ:1.5822388116518658
NB_PESQ:2.4539393933614093

other:
![DNS2020_result](https://user-images.githubusercontent.com/3354428/189028805-bde6529a-5a1a-42d7-8aec-5ad99a04bb54.png)

DataSet: https://github.com/microsoft/DNS-Challenge/tree/interspeech2020/master/datasets/test_set/synthetic/no_reverb

compute metrics scripts:
```

# -*- coding: utf-8 -*-
import glob
import sys
import os
import librosa
import numpy as np
from tqdm import tqdm
from joblib import Parallel, delayed
import argparse
from pystoi.stoi import stoi
from pesq import pesq
from pypesq import pesq as nb_pesq
import torch

EPS = np.finfo(np.float32).eps

REGISTERED_METRICS = ["SI_SDR", "STOI", "WB_PESQ", "NB_PESQ"]


def SI_SDR(noisy, clean):
    if torch.is_tensor(noisy) and torch.is_tensor(clean):
        clean_energy = torch.sum(clean ** 2, dim=-1, keepdim=True)

        optimal_scaling = torch.sum(clean * noisy, dim=-1, keepdim=True) / (clean_energy + EPS)

        projection = optimal_scaling * clean

        noise = noisy - projection

        ratio = torch.sum(projection ** 2, dim=-1) / (torch.sum(noise ** 2, dim=-1) + EPS)
        return 10 * torch.log10(ratio + EPS)
    else:
        noisy, clean = np.broadcast_arrays(noisy, clean)
        clean_energy = np.sum(clean ** 2, axis=-1, keepdims=True)

        optimal_scaling = np.sum(clean * noisy, axis=-1, keepdims=True) / (clean_energy + EPS)

        projection = optimal_scaling * clean

        noise = noisy - projection

        ratio = np.sum(projection ** 2, axis=-1) / (np.sum(noise ** 2, axis=-1) + EPS)
        return 10 * np.log10(ratio + EPS)


def STOI(noisy, clean, sr=16000):
    return stoi(clean, noisy, sr, extended=False)


def WB_PESQ(noisy, clean, sr=16000):
    return pesq(sr, clean, noisy, "wb")


def NB_PESQ(noisy, clean, sr=16000):
    return nb_pesq(clean, noisy, sr)

def calculate_metric(noisy_file, clean_file, sr=16000, metric_type="STOI", pre_load=False):
    # get noisy, clean
    clean_file = os.path.abspath(clean_file)
    noisy_file = os.path.abspath(noisy_file)
    if pre_load == False:
        noisy, _ = librosa.load(noisy_file, sr=sr)
        clean, _ = librosa.load(clean_file, sr=sr)
    else:
        noisy = noisy_file
        clean = clean_file
    assert len(noisy) == len(clean)

    # get metric score
    if metric_type in ["SI_SDR"]:
        return SI_SDR(noisy, clean)
    elif metric_type in ["STOI"]:
        return STOI(noisy, clean, sr=sr)
    elif metric_type in ["WB_PESQ"]:
        return WB_PESQ(noisy, clean)
    elif metric_type in ["NB_PESQ"]:
        return NB_PESQ(noisy, clean)


def compute_metric(noisy_files, clean_files, metrics, n_folds=1, n_jobs=8, pre_load=False):
    for metric_type, _ in metrics.items():
        assert metric_type in REGISTERED_METRICS

        split_num = len(noisy_files) // n_folds
        score = []
        for n in range(n_folds):
            metric_score = Parallel(n_jobs=n_jobs)(
                delayed(calculate_metric)(
                    noisy_file,
                    clean_file,
                    sr=16000,
                    metric_type=metric_type,
                    pre_load=pre_load,
                )
                for noisy_file, clean_file in tqdm(
                    zip(
                        noisy_files[n * split_num : (n + 1) * split_num],
                        clean_files[n * split_num : (n + 1) * split_num],
                    )
                )
            )
            score.append(np.mean(metric_score))
        metrics[metric_type] = np.mean(score)

def get_fileid_n(filename: str):
    filename = os.path.splitext(filename)[0]
    file_id = filename.split("_fileid_")[1].zfill(10)
    return file_id


def collect_wav_files(input_dir, key=get_fileid_n):
    return sorted(glob.glob(os.path.join(input_dir, "*.wav")), key=key)



if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="""
            Deep Noise Suppression Model Score Evaluation
            """,
            formatter_class=argparse.ArgumentDefaultsHelpFormatter
            )
    
    # positional arguments
    parser.add_argument("ref_dir", metavar="reference_dir",
                        type=str,
                        help="""
                        reference directory containing clean speech files
                        """)
    
    parser.add_argument("est_dir", metavar="estimates_dir",
                        type=str,
                        help="""
                        estimates directory containing predicted speech files
                        """)
    
    args = parser.parse_args()
    # check if directories exist
    if not os.path.isdir(args.ref_dir):
        raise FileNotFoundError(
                f"Reference directory not found: {args.ref_dir}"
                )

    if not os.path.isdir(args.est_dir):
        raise FileNotFoundError(
                f"Estimates directory not found: {args.est_dir}"
                )

    clean_files = collect_wav_files(args.ref_dir)
    
    noisy_files = collect_wav_files(args.est_dir)

    metrics = {
        "SI_SDR": [],
        "STOI": [],
        "WB_PESQ": [],
        "NB_PESQ": [],
    }
    
    compute_metric(
        noisy_files,
        clean_files,
        metrics,
        n_folds=1,
        n_jobs=8,
        pre_load=False,
    )
    
    for metric_type, _ in metrics.items():
        print(f"{metric_type}:{metrics[metric_type]}")
```

**Comment by Rikorose** [comment]
Hi there, since SISDR is negative and enhanced STOI is smaller than noisy STOI, I would assume, that there is a bug in the evaluation implementation. It could be an temporal offset, e.g. since DeepFilterNet uses a real-time STFT implementation that introduces an offset. One option you could try is to run `deepFilter` with `-D` option which compensates the STFT delay.

**Comment by zuowanbushiwo** [comment]
@Rikorose 
Which window function are you using in `frame_analysis /frame_synthesis` ? I compare ` df.loss.Stft` compared  and 
`df.analysis`, and their numerically  are very different.  

```
    sr, fft, hop, nb_bands = 48000, 960, 480, 32
    df = DF(sr, fft, hop, nb_bands)
    x, sr = librosa.load(librosa.ex("trumpet"), sr=sr)
    sf.write("in.wav", x.T, sr)
    x = x.reshape(1, -1)
    x_t = torch.from_numpy(x)
    stft = Stft(n_fft=960, hop=480)  # from df.loss.Stft
    x_stft = stft(x_t).permute(0, 2, 1)
    x_stft_norm = torch.norm(x_stft)
    x = df.analysis(x)
    x_norm = np.linalg.norm(x)
    print(f"x_norm:{x_norm},x_stft_norm:{x_stft_norm}")
```
`output: x_norm:0.884440541267395, ,x_stft_norm:23.73641586303711` and their shape   are also different .

How to implement frame_synthesis/ frame_synthesis  with pytorch?

**Comment by Rikorose** [comment]
Yes the normalization factor is different between the torch implementation. This should be a constant factor though.

**Comment by zuowanbushiwo** [comment]
@Rikorose 
Thanks，So  what window function are you using in frame_analysis /frame_synthesis?  I compared common windows, none of them look like, `matlab` command:

`wvtool(hamming(960),hann(960),gausswin(960),df_win)` //  df_win is DeepFilter  frame_analysis /frame_synthesis windows

<img width="932" alt="image" src="https://user-images.githubusercontent.com/3354428/189253254-56b555b2-81a9-4fb6-912e-1e1416ac6a13.png">

**Comment by Rikorose** [comment]
I use the vorbis window similar to RNNoise 
![snippet](https://user-images.githubusercontent.com/16517898/189286059-5f2c5613-a32e-4564-9aae-d90fed4b68bc.png)

**Comment by zuowanbushiwo** [comment]
@Rikorose 
Thank you very much, you are so friendly！

**Comment by Rikorose** [comment]
You're welcome!
